File: assets/code_box/bk_2580.c
#include <stdio.h>
#include <stdlib.h>

int blank_count = 0 ;
int sudoku[15][15] = { 0, };
int blank_idx[100][5] = { 0, };
// 반례 참고 https://cryptosalamander.tistory.com/59
// 444444hello3hello2hello2hellohellohellohellohellohello43324
// int is_in(int num, int arr[] ,int arr_leng)
// {
//     for(int i = 0; i< arr_leng; i++)
//     {
//         if(num == arr[i]) return 0;
//     }
//     return 1;
// }

void back(int n,int fill_blanks[])
{
    // int k =0;
    // for(int i = 0; i<9; i++)
    // {
    //     for(int j = 0; j<9; j++)
    //     {
    //         if(sudoku[i][j] == 0 && j < n)
    //         {
    //             printf("%d",fill_blanks[k++]);
    //         }
    //         else
    //         {
    //             printf("%d",sudoku[i][j]);
    //         }
    //         if(j < 8) printf(" ");
    //     }
    //     printf("\n");
    // }
    // printf("\n");
    if(n == blank_count)
    {
        int k = 0;
        for(int i = 0; i<9; i++)
        {
            for(int j = 0; j<9; j++)
            {
                if(sudoku[i][j] == 0)
                {
                    printf("%d",fill_blanks[k++]);
                }
                else
                {
                    printf("%d",sudoku[i][j]);
                }
                if(j < 8) printf(" ");
            }
            printf("\n");
        }
        exit(0);
    }
    else
    {
        for( int i = 1; i< 10; i++)
        {
            int is_fill = 1;
            int smallcube_row = blank_idx[n][0] / 3;
            int smallcube_col = blank_idx[n][1] / 3;
            for(int j = 0; j < 9; j++)
            {
                // if(i == 7) printf("hello");
                if(sudoku[j][blank_idx[n][1]] == i || sudoku[blank_idx[n][0]][j] == i )
                {
                    // printf("hello!\n");
                    is_fill = 0;
                    // printf("4");
                    break;
                }
                else if( j < 3)
                {
                    for(int p = smallcube_col* 3 ; p < smallcube_col* 3 + 3; p++)
                    {
                        
                        // if( n ==3 && blank_idx[n][0] == 2 && blank_idx[n][1] == 0)
                        // {
                        //     printf("sudoku [%d] [%d] = %d",smallcube_row * 3 + j,p,sudoku[smallcube_row * 3 + j][p]);
                        // }
                        if(sudoku[smallcube_row * 3 + j][p] == 0)
                        {
                            for(int u = 0; u < n;u++)
                            {
                                if(blank_idx[u][0] == smallcube_row * 3 + j && blank_idx[u][1] ==p )
                                {
                                    if(fill_blanks[u] == i)
                                    {
                                        is_fill = 0;
                                        // printf("1");
                                        break;
                                    }
                                }
                            }
                            if(!is_fill)
                            {
                                // printf("2");
                                break;
                            }
                        }
                        else if(sudoku[smallcube_row * 3 + j][p] == i )
                        {

                            // printf("sudoku[%d][%d] = %d",smallcube_row * 3 + j,p,i);

                            is_fill = 0;
                            // printf("3");
                            break;                        
                        }

                    }

                }
            }
            if(is_fill)
            {
                fill_blanks[n] = i;
                back(n+1,fill_blanks);
                fill_blanks[n] = 0;
            }
        }
    }
    
}


int main()
{

    for(int i = 0; i<9; i++)
    {
        scanf("%d %d %d %d %d %d %d %d %d", &sudoku[i][0], &sudoku[i][1], &sudoku[i][2], &sudoku[i][3], &sudoku[i][4], &sudoku[i][5], &sudoku[i][6], &sudoku[i][7], &sudoku[i][8]);
        for (int j = 0; j < 9; j ++)
        {
            if(sudoku[i][j] == 0)
            {
                blank_idx[blank_count][0] = i;
                blank_idx[blank_count][1] = j;
                blank_count++;
            }
        }
    }
    int fill_blanks[100] = {0,};
    back(0,fill_blanks);

}

File: assets/code_box/bk_2580.py
import copy
sudoku = []
blank_idx = []
for i in range(9):
    sudoku_row = []
    sudoku_row = list(map(int, input().split()))
    sudoku.append(sudoku_row)
    r = -1
    for j in range(sudoku_row.count(0)):
        # print(sudoku_row.count(0))
        row = sudoku_row[r+1:]
        blank_idx.append([i,r + 1 + row.index(0)])
        # print(i , " ", r + 1 + row.index(0))
        r = row.index(0)

def back(n,fill_blank):
    fill = copy.deepcopy(fill_blank)
    global total_blanks,blank_idx,sudoku
    if(n == total_blanks):
        for i in range(9):
            for j in range(9):
                if(sudoku[i][j] == 0):
                    print(fill[0],end="")
                    if(len(fill) > 1):
                        fill = fill[1:]
                else:
                    print(sudoku[i][j],end= "")
                if(j < 8):
                    print("",end=" ")
            print("")
    else :
        for i in range(1,10):
            is_fill = True
            if(not i in sudoku[blank_idx[n][0]]):
                smallcube_row = blank_idx[n][0] // 3
                smallcube_col = blank_idx[n][1] // 3
                for j in range(9):
                    if(sudoku[j][blank_idx[n][1]] == i):
                        is_fill = False
                        break
                    if( j < 3):
                        if (i in sudoku[smallcube_row * 3 + j][smallcube_col * 3 : smallcube_col * 3 + 3]):
                            is_fill = False
                            break
                if(is_fill):
                    fill.append(i)
                    back(n+1,fill)
                

                
        


total_blanks = len(blank_idx)
# print(total_blanks)
back(0,[])